File: app/src/main/java/forpdateam/ru/forpda/fragments/auth/AuthFragment.java
package forpdateam.ru.forpda.fragments.auth;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.github.rahatarmanahmed.cpv.CircularProgressView;
import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.TabManager;
import forpdateam.ru.forpda.api.Utils;
import forpdateam.ru.forpda.api.auth.models.AuthForm;
import forpdateam.ru.forpda.api.profile.models.ProfileModel;
import forpdateam.ru.forpda.client.ClientHelper;
import forpdateam.ru.forpda.fragments.TabFragment;
import forpdateam.ru.forpda.fragments.news.main.NewsMainFragment;
import forpdateam.ru.forpda.rxapi.RxApi;
import forpdateam.ru.forpda.utils.EmptyActivity;
import forpdateam.ru.forpda.utils.SimpleTextWatcher;
import forpdateam.ru.forpda.utils.rx.Subscriber;
import forpdateam.ru.forpda.views.drawers.Drawers;

/**
 * Created by radiationx on 29.07.16.
 */
public class AuthFragment extends TabFragment {
    private EditText nick, password, captcha;
    private ImageView captchaImage, avatar;
    private AuthForm authForm;
    private Button sendButton;
    private Button skipButton;
    private ProgressBar loginProgress, captchaProgress;
    private CheckBox hiddenAuth;

    private LinearLayout mainForm;
    private RelativeLayout complete;
    private TextView completeText;
    private CircularProgressView progressView;
    private Subscriber<AuthForm> mainSubscriber = new Subscriber<>(this);
    private Subscriber<Boolean> loginSubscriber = new Subscriber<>(this);
    private Subscriber<ProfileModel> profileSubscriber = new Subscriber<>(this);

    public AuthFragment() {
        configuration.setAlone(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        baseInflateFragment(inflater, R.layout.fragment_auth);
        nick = (EditText) findViewById(R.id.auth_login);
        password = (EditText) findViewById(R.id.auth_password);
        captcha = (EditText) findViewById(R.id.auth_captcha);
        captchaImage = (ImageView) findViewById(R.id.captchaImage);
        captchaProgress = (ProgressBar) findViewById(R.id.captcha_progress);
        avatar = (ImageView) findViewById(R.id.auth_avatar);
        mainForm = (LinearLayout) findViewById(R.id.auth_main_form);
        complete = (RelativeLayout) findViewById(R.id.auth_complete);
        completeText = (TextView) findViewById(R.id.auth_complete_text);
        progressView = (CircularProgressView) findViewById(R.id.auth_progress);
        loginProgress = (ProgressBar) findViewById(R.id.login_progress);
        hiddenAuth = (CheckBox) findViewById(R.id.auth_hidden);
        sendButton = (Button) findViewById(R.id.auth_send);
        skipButton = (Button) findViewById(R.id.auth_skip);
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewsReady();
        setListsBackground();
        skipButton.setOnClickListener(v -> {
            new AlertDialog.Builder(getContext())
                    .setMessage("Без авторизации будут недоступны некоторые функции приложения.")
                    .setPositiveButton(R.string.ok, (dialog, which) -> {
                        Drawers drawers = getMainActivity().getDrawers();
                        drawers.selectMenuItem(NewsMainFragment.class);
                        TabManager.getInstance().remove(AuthFragment.this);
                    })
                    .setNegativeButton(R.string.cancel, null)
                    .show();
        });
        appBarLayout.setVisibility(View.GONE);
        notifyDot.setVisibility(View.GONE);
        sendButton.setOnClickListener(v -> tryLogin());
        LoginTextWatcher loginTextWatcher = new LoginTextWatcher();
        nick.addTextChangedListener(loginTextWatcher);
        password.addTextChangedListener(loginTextWatcher);
        captcha.addTextChangedListener(loginTextWatcher);
        fragmentContainer.setFitsSystemWindows(true);
        fragmentContent.setFitsSystemWindows(true);
    }


    @Override
    protected void updateNotifyDot() {
        //Чтобы не показывалась точка.
    }

    @Override
    public boolean loadData() {
        if (!super.loadData()) {
            return false;
        }
        mainSubscriber.subscribe(RxApi.Auth().getForm(), this::onLoadForm, new AuthForm(), view1 -> loadData());
        return true;
    }

    private void onLoadForm(AuthForm authForm) {
        if (authForm.getBody() == null) return;
        this.authForm = authForm;
        captchaProgress.setVisibility(View.VISIBLE);
        captchaImage.setVisibility(View.GONE);
        ImageLoader.getInstance().displayImage(authForm.getCaptchaImageUrl(), captchaImage, new SimpleImageLoadingListener() {
            @Override
            public void onLoadingComplete(String imageUri, View view, Bitmap loadedImage) {
                captchaProgress.setVisibility(View.GONE);
                captchaImage.setVisibility(View.VISIBLE);
            }
        });
    }

    private void tryLogin() {
        if (authForm == null) return;
        authForm.setCaptcha(captcha.getText().toString());
        authForm.setNick(nick.getText().toString());
        authForm.setPassword(password.getText().toString());
        authForm.setHidden(hiddenAuth.isChecked());
        loginProgress.setVisibility(View.VISIBLE);
        sendButton.setVisibility(View.INVISIBLE);
        hidePopupWindows();
        loginSubscriber.subscribe(RxApi.Auth().login(authForm), this::showLoginResult, false, view1 -> loadData());
        //showLoginResult(false);
    }

    private void showLoginResult(boolean b) {
        Log.d(AuthFragment.class.getSimpleName(), "showLoginResult " + b);
        if (b) {
            loadProfile();
        } else {
            loadData();
            captcha.setText("");
            password.setText("");
        }
        loginProgress.setVisibility(View.GONE);
        sendButton.setVisibility(View.VISIBLE);
    }

    public void loadProfile() {

        AlphaAnimation animation = new AlphaAnimation(1.0f, 0.0f);
        animation.setDuration(225);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                mainForm.setVisibility(View.GONE);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        mainForm.startAnimation(animation);
        complete.setVisibility(View.VISIBLE);
        AlphaAnimation animation1 = new AlphaAnimation(0.0f, 1.0f);
        animation1.setDuration(375);
        complete.startAnimation(animation1);

        profileSubscriber.subscribe(RxApi.Profile().getProfile("https://4pda.ru/forum/index.php?showuser=".concat(Integer.toString(ClientHelper.getUserId()))), this::onProfileLoad, new ProfileModel());
    }

    private void onProfileLoad(ProfileModel profile) {
        if (EmptyActivity.empty(profile.getNick())) {
            loginProgress.setVisibility(View.GONE);
            // Надо записать иначе будет каждый раз просить авторизацию
            App.get().getPreferences().edit().putString("auth.user.nick", profile.getNick()).apply();
            EmptyActivity.start(getActivity(), profile.getNick());
            getMainActivity().finish();
            return;
        }
        App.get().getPreferences().edit().putString("auth.user.nick", profile.getNick()).apply();
        ImageLoader.getInstance().displayImage(profile.getAvatar(), avatar);
        completeText.setText(Utils.spannedFromHtml(String.format("%s, <b>%s</b>!", getString(R.string.auth_hello), profile.getNick())));
        completeText.setVisibility(View.VISIBLE);
        AlphaAnimation animation = new AlphaAnimation(0.0f, 1.0f);
        animation.setDuration(1000);
        completeText.startAnimation(animation);
        AlphaAnimation animation1 = new AlphaAnimation(1.0f, 0.0f);
        animation1.setDuration(225);
        animation1.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                progressView.setVisibility(View.GONE);
                progressView.stopAnimation();
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        progressView.startAnimation(animation1);
        new Handler().postDelayed(() -> {
            ClientHelper.get().notifyAuthChanged(ClientHelper.AUTH_STATE_LOGIN);
            new Handler().postDelayed(() -> {
                TabManager.getInstance().remove(AuthFragment.this);
            }, 500);
        }, 2000);
    }

    private class LoginTextWatcher extends SimpleTextWatcher {
        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            if (!nick.getText().toString().isEmpty() && !password.getText().toString().isEmpty() && captcha.getText().toString().length() == 4) {
                if (!sendButton.isEnabled())
                    sendButton.setEnabled(true);
            } else {
                if (sendButton.isEnabled())
                    sendButton.setEnabled(false);
            }
        }
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/auth/AuthFragment.kt
package forpdateam.ru.forpda.ui.fragments.auth

import android.graphics.Bitmap
import android.os.Bundle
import android.view.View
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.inputmethod.EditorInfo
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.appcompat.app.AlertDialog
import com.github.rahatarmanahmed.cpv.CircularProgressView
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.common.simple.SimpleAnimationListener
import forpdateam.ru.forpda.common.simple.SimpleTextWatcher
import forpdateam.ru.forpda.databinding.FragmentAuthBinding
import forpdateam.ru.forpda.entity.remote.auth.AuthCaptcha
import forpdateam.ru.forpda.entity.remote.auth.AuthForm
import forpdateam.ru.forpda.entity.remote.profile.ProfileModel
import forpdateam.ru.forpda.extensions.quillMoxyPresenter
import forpdateam.ru.forpda.model.data.remote.api.ApiUtils
import forpdateam.ru.forpda.presentation.auth.AuthPresenter
import forpdateam.ru.forpda.presentation.auth.AuthView
import forpdateam.ru.forpda.ui.fragments.TabFragment
import forpdateam.ru.forpda.ui.fragments.tabBinding

/**
 * Created by radiationx on 29.07.16.
 */
class AuthFragment : TabFragment(R.layout.fragment_auth), AuthView {

    private val binding by tabBinding(FragmentAuthBinding::bind)

    private val nick: EditText
        get() = binding.authLogin
    private val password: EditText
        get() = binding.authPassword
    private val captcha: EditText
        get() = binding.authCaptcha
    private val captchaImage: ImageView
        get() = binding.captchaImage
    private val avatar: ImageView
        get() = binding.authAvatar
    private val sendButton: Button
        get() = binding.authSend
    private val skipButton: Button
        get() = binding.authSkip
    private val regButton: Button
        get() = binding.authReg
    private val loginProgress: ProgressBar
        get() = binding.loginProgress
    private val captchaProgress: ProgressBar
        get() = binding.captchaProgress
    private val hiddenAuth: CheckBox
        get() = binding.authHidden

    private val mainForm: LinearLayout
        get() = binding.authMainForm
    private val complete: RelativeLayout
        get() = binding.authComplete
    private val completeText: TextView
        get() = binding.authCompleteText
    private val progressView: CircularProgressView
        get() = binding.authProgress

    private val authTopButtons: FrameLayout
        get() = binding.authTopButtons


    private val loginTextWatcher = object : SimpleTextWatcher() {
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            updateForm()
        }
    }

    private val presenter by quillMoxyPresenter<AuthPresenter>()

    init {
        configuration.defaultTitle = getString(R.string.fragment_title_auth)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setListsBackground()
        skipButton.setOnClickListener { v ->
            AlertDialog.Builder(requireContext())
                .setMessage("Без авторизации будут недоступны некоторые функции приложения.")
                .setPositiveButton(R.string.ok) { dialog, which -> presenter.onClickSkip() }
                .setNegativeButton(R.string.cancel, null)
                .show()
        }
        regButton.setOnClickListener {
            AlertDialog.Builder(requireContext())
                .setMessage("Процесс регистрации включает в себя множество шагов, поэтому рекомендуем зарегистрироваться через браузер.")
                .setPositiveButton(R.string.ok) { _, _ -> presenter.onRegistrationClick() }
                .setNegativeButton(R.string.cancel, null)
                .show()
        }
        appBarLayout.visibility = View.GONE
        sendButton.setOnClickListener { v -> tryLogin() }
        nick.addTextChangedListener(loginTextWatcher)
        password.addTextChangedListener(loginTextWatcher)
        captcha.addTextChangedListener(loginTextWatcher)
        fragmentContainer.fitsSystemWindows = true
        fragmentContent.fitsSystemWindows = true

        hiddenAuth.setOnCheckedChangeListener { buttonView, isChecked ->
            updateForm()
        }

        captcha.setOnEditorActionListener(OnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                if (sendButton.isEnabled) {
                    tryLogin()
                }
                return@OnEditorActionListener true
            }
            false
        })
    }

    override fun setSendEnabled(isEnabled: Boolean) {
        sendButton.isEnabled = isEnabled
    }

    override fun setSendRefreshing(isRefreshing: Boolean) {
        if (isRefreshing) {
            loginProgress.visibility = View.VISIBLE
            sendButton.visibility = View.INVISIBLE
        } else {
            loginProgress.visibility = View.INVISIBLE
            sendButton.visibility = View.VISIBLE
        }
    }

    override fun onCaptchaLoaded(authCaptcha: AuthCaptcha) {
        captchaImage.visibility = View.GONE
        captchaProgress.visibility = View.VISIBLE
        ImageLoader.getInstance().displayImage(
            authCaptcha.captchaImageUrl,
            captchaImage,
            object : SimpleImageLoadingListener() {
                override fun onLoadingComplete(
                    imageUri: String?,
                    view: View?,
                    loadedImage: Bitmap?
                ) {
                    captchaImage.visibility = View.VISIBLE
                    captchaProgress.visibility = View.GONE
                }
            })
    }

    override fun onFormChanged(authForm: AuthForm) {
        nick.setText(authForm.nick)
        password.setText(authForm.password)
        captcha.setText(authForm.captcha)
        hiddenAuth.isChecked = authForm.isHidden
    }

    private fun tryLogin() {
        hideKeyboard()
        presenter.signIn()
    }

    override fun onSuccessAuth() {
        mainForm.startAnimation(AlphaAnimation(1.0f, 0.0f).apply {
            duration = 225
            setAnimationListener(object : SimpleAnimationListener() {
                override fun onAnimationEnd(animation: Animation) {
                    mainForm.visibility = View.GONE
                }
            })
        })
        authTopButtons.startAnimation(AlphaAnimation(1.0f, 0.0f).apply {
            duration = 225
            setAnimationListener(object : SimpleAnimationListener() {
                override fun onAnimationEnd(animation: Animation) {
                    authTopButtons?.visibility = View.GONE
                }
            })
        })
        complete.visibility = View.VISIBLE
        complete.startAnimation(AlphaAnimation(0.0f, 1.0f).apply {
            duration = 375
        })
    }

    override fun showProfile(profile: ProfileModel) {
        ImageLoader.getInstance().displayImage(profile.user.avatar, avatar)
        completeText.text = ApiUtils.spannedFromHtml("${getString(R.string.auth_hello)}, <b>${profile.user.nick}</b>!")
        completeText.visibility = View.VISIBLE

        completeText.startAnimation(AlphaAnimation(0.0f, 1.0f).apply {
            duration = 1000
        })

        progressView.startAnimation(AlphaAnimation(1.0f, 0.0f).apply {
            duration = 225
            setAnimationListener(object : SimpleAnimationListener() {
                override fun onAnimationEnd(animation: Animation) {
                    progressView.visibility = View.GONE
                    progressView.stopAnimation()
                }
            })
        })
    }

    private fun updateForm() {
        presenter.updateForm(
            AuthForm(
                nick = nick.text?.toString().orEmpty(),
                password = password.text?.toString().orEmpty(),
                captcha = captcha.text?.toString().orEmpty(),
                isHidden = hiddenAuth.isChecked
            )
        )
    }
}
